renderer/FrameBuffer.cpp

```diff
@@ -464,3 +464,221 @@ void FB_TogglePrimary( bool on ) {
+	for( auto it : frameBuffers.allFrameBuffers ) {
+		delete it;
+	}
+	frameBuffers.allFrameBuffers.clear();
+
+	frameBuffers.primary = nullptr;
+	frameBuffers.resolve = nullptr;
+	frameBuffers.lightgem = nullptr;
+	frameBuffers.backBuffer = nullptr;
+	frameBuffers.finalOutput = nullptr;
+	frameBuffers.currentDraw = nullptr;
+	frameBuffers.currentRead = nullptr;
+}
+
+FrameBuffer::FrameBuffer( GLuint fbo, GLuint width, GLuint height, int msaa )
+	: fbo( fbo ), colorBuffer( 0 ), depthStencilBuffer( 0 ), width( width ), height( height ), msaa( msaa ) {
+}
+
+
+FrameBuffer::~FrameBuffer() {
+	if( fbo != 0 ) {
+		qglDeleteFramebuffers( 1, &fbo );
+	}
+	if( colorBuffer != 0 ) {
+		qglDeleteRenderbuffers( 1, &colorBuffer );
+	}
+	if( depthStencilBuffer != 0 ) {
+		qglDeleteRenderbuffers( 1, &depthStencilBuffer );
+	}
+}
+
+FrameBuffer * FrameBuffer::Create( GLuint width, GLuint height, int msaa ) {
+	GLuint fbo = 0;
+	qglGenFramebuffers( 1, &fbo );
+	if( fbo == 0 ) {
+		common->FatalError( "Failed to allocate framebuffer object" );
+	}
+
+	FrameBuffer *frameBuffer = new FrameBuffer( fbo, width, height, msaa );
+	frameBuffers.allFrameBuffers.push_back( frameBuffer );
+	common->Printf( "FBO created: %dx%d, AA = %d\n", width, height, msaa );
+	return frameBuffer;
+}
+
+FrameBuffer * FrameBuffer::BackBuffer() {
+	if( frameBuffers.backBuffer == 0 ) {
+		frameBuffers.backBuffer = new FrameBuffer( 0, glConfig.vidWidth, glConfig.vidHeight, 0 );
+		frameBuffers.allFrameBuffers.push_back( frameBuffers.backBuffer );
+		common->Printf( "Backbuffer dimensions: %dx%d\n", frameBuffers.backBuffer->width, frameBuffers.backBuffer->height );
+	}
+	return frameBuffers.backBuffer;
+}
+
+void FrameBuffer::CreateColorBuffer() {
+	if( colorBuffer == 0 ) {
+		qglGenRenderbuffers( 1, &colorBuffer );
+		if( colorBuffer == 0 ) {
+			common->FatalError( "Failed to allocate color render buffer" );
+		}
+	}
+	qglBindRenderbuffer( GL_RENDERBUFFER, colorBuffer );
+	if( msaa > 1 ) {
+		qglRenderbufferStorageMultisample( GL_RENDERBUFFER, msaa, GL_RGBA, width, height );
+	} else {
+		qglRenderbufferStorage( GL_RENDERBUFFER, GL_RGBA, width, height );
+	}
+
+	Bind();
+	qglFramebufferRenderbuffer( GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER, colorBuffer );
+}
+
+void FrameBuffer::CreateDepthStencilBuffer() {
+	if( depthStencilBuffer == 0 ) {
+		qglGenRenderbuffers( 1, &depthStencilBuffer );
+		if( depthStencilBuffer == 0 ) {
+			common->FatalError( "Failed to allocate depth/stencil render buffer" );
+		}
+	}
+	qglBindRenderbuffer( GL_RENDERBUFFER, depthStencilBuffer );
+	if( msaa > 1 ) {
+		qglRenderbufferStorageMultisample( GL_RENDERBUFFER, msaa, GL_DEPTH24_STENCIL8, width, height );
+	} else {
+		qglRenderbufferStorage( GL_RENDERBUFFER, GL_DEPTH24_STENCIL8, width, height );
+	}
+
+	Bind();
+	qglFramebufferRenderbuffer( GL_FRAMEBUFFER, GL_DEPTH_STENCIL_ATTACHMENT, GL_RENDERBUFFER, depthStencilBuffer );
+}
+
+void FrameBuffer::AddColorTexture( idImage *colorTexture ) {
```

Comment:
Is there any way to ensure that images does not die before FBO is deleted?
Renderbuffers are owned by FBO, so they won't die before FBO dies.
But it might happen so that idImage (or its gl texture object) is deleted before FBO is, which can cause serious issues.

Reply:
It would render the FBO invalid, definitely. However, I don't see a clear way to prevent that. You would need a different ownership model for idImage to stop someone from deleting it before the FBO is released.
We could perhaps add a kind of marker to a texture, but that doesn't strike me as particularly elegant.

Reply:
Isn't there a global manager for all `idImage`-s ? I guess `idImageManager *globalImages` works like this. All images are allocated via `idImageManager::AllocImage` method and deallocated via `idImage::PurgeImage` method.

In this case it is possible to check if an image is attached to FBO when it is purged, and raise an error in this case.